External package detection for compilers

becker33 opened the pull request:
This PR creates shared infrastructure for compiler packages to implement the detailed search capabilities from the `spack compiler find` command for the `spack external find` command.

This PR updates the `aocc`, `apple-clang`, `arm`, `gcc`, and `nag` packages to use the new functionality. PRs for other compilers to come.

With this PR, `spack compiler find` can be replaced with `spack external find --tag compiler` as long as only the five covered compilers are relevant.

Once the rest of the packages are updated, future work will replace the `spack compiler find` command with an alias to `spack external find --tag compiler`.

becker33 committed:
CompilerPackage mixin for compiler detection

Adapted gcc package but not other compiler packages yet

becker33 committed:
update nag

becker33 committed:
update aocc

becker33 committed:
update apple-clang

becker33 committed:
add package for arm compiler

becker33 committed:
add testing for compiler detection through externals

becker33 committed:
cleanup rebase artifacts

becker33 committed:
fix apple-clang audits to match new version detection

becker33 committed:
fix gcc compiler prefixes

becker33 committed:
fixup packages -> compilers translation to fill in null entries

becker33 committed:
update docs to match

becker33 committed:
fixup test that relied on old behavior

becker33 committed:
missing newline for windows mock executables

becker33 committed:
debug for windows

more debug prints

more debuggin

more debugging

more debuggin

more debugging

more debugging

fix executables to be str not re.compile

remove debug prints

fix windows executable suffixes in test

fixup d11d924429

becker33 committed:
more complex logic to match up exe with lang

alalazo commented on var/spack/repos/builtin/packages/apple-clang/detection_test.yaml:
```diff
@@ -4,32 +4,20 @@ paths:
       - executables:
           - "bin/clang"
           - "bin/clang++"
-        script: |
-          echo "Apple clang version 11.0.0 (clang-1100.0.33.8)"
-          echo "Target: x86_64-apple-darwin19.5.0"
-          echo "Thread model: posix"
-          echo "InstalledDir: /Library/Developer/CommandLineTools/usr/bin"
```
I guess this is a temporary change, right?

becker33 replied:
No, that's a change because we changed what argument we pass to `apple-clang` to detect the version. `apple-clang -dumpversion` outputs only the version string itself, rather than all of this information that we don't need for detection.

alalazo replied:
How do we distinguish between Apple Clang and LLVM Clang then?

becker33 replied:
Ugh yep that's annoying. Fixed.

becker33 committed:
make conversion robust to missing package property

becker33 committed:
improve language to exe mapping

becker33 committed:
rewrite mapping from languages to exes

becker33 committed:
update apple clang to differentiate from clang

becker33 committed:
update fj compiler

alalazo commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -537,70 +546,26 @@ def filter_detected_exes(cls, prefix, exes_in_prefix):
+        assert "paths" in extra_attributes, msg
 
-        compilers = extra_attributes["compilers"]
+        paths = extra_attributes["paths"]
```
Can we keep using `compilers` and fix the couple of lines in `compilers/__init__.py`?

becker33 replied:
Yeah I can make that happen. I'll also need to update docs from https://spack.readthedocs.io/en/latest/getting_started.html#manual-compiler-configuration

becker33 replied:
I've gone back and forth on this question at least twice now, glad to have someone else resolve it for me

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,126 @@
+import spack.package_base
+
+
+class CompilerPackage(spack.package_base.PackageBase):
```
Needs a docstring, to explain the purpose of the mixin

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,126 @@
+
+
+class CompilerPackage(spack.package_base.PackageBase):
+    # metadata identifying this as a compiler for lmod
+    family = "compiler"
```
I can't find where this is used... can you please point me to the logic using it? :thinking:

becker33 replied:
The logic is used in the lmod generation IIRC, but it's not new in the PR. This PR just ensures it is appropriately applied to every package that is a compiler.

alalazo replied:
Yeah, what I'm saying is:
```console
$ grep -R -A3 \.family lib/spack/spack/modules/*.py
lib/spack/spack/modules/lmod.py:        # If it is in the list of supported compilers family -> compiler
lib/spack/spack/modules/lmod.py-        if self.spec.name in spack.compilers.supported_compilers():
lib/spack/spack/modules/lmod.py-            provides["compiler"] = spack.spec.CompilerSpec(self.spec.format("{name}{@versions}"))
lib/spack/spack/modules/lmod.py-        elif self.spec.name in spack.compilers.package_name_to_compiler_name:
--
lib/spack/spack/modules/lmod.py:                [str(self.spec.platform), str(self.spec.os), str(self.spec.target.family)]
lib/spack/spack/modules/lmod.py-            )
lib/spack/spack/modules/lmod.py-            return os.path.join(self.dirname(), arch_folder)
lib/spack/spack/modules/lmod.py-        return self.dirname()
```
so it doesn't seem used anymore.

alalazo replied:
I think we can remove this. No need to propagate outdated information:
```suggestion
```

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,126 @@
+    #: Regex used to extract version from compiler's output
+    version_regex = "(.*)"
+
+    #: Platform matcher for Platform objects supported by compiler
+    is_supported_on_platform = lambda x: True
```
Let's not use lambdas at class level. We should define a method and make it clear if it is a static method (I assume `x` is the platform)

becker33 replied:
I agree in principle, but I went and implemented it and it's really ugly.

The way I see it, there are 4 options:

1. `lambda x: True`
2. Define a static method in class scope
```
@staticmethod
def true():
    return True

is_supported_on_platform = true
```
3. Do the exact same thing with a method in module scope instead
4. `is_supported_on_platform = object`, because `object` is callable and always returns True.

I've implemented option 3, but I think I prefer the lambda over that (and that's what we had previously in the `spack.compiler.Compiler` class).

alalazo replied:
I think this function is not used, will comment in the next review and close this.

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,126 @@
+    #: Regex used to extract version from compiler's output
+    version_regex = "(.*)"
+
+    #: Platform matcher for Platform objects supported by compiler
+    is_supported_on_platform = lambda x: True
+
+    #: Static definition of languages supported by this class
+    compiler_languages = ["c", "cxx", "fortran"]
+
+    #: Dynamic definition of languages supported by this package
+    @property
+    def supported_languages(self):
+        return self.compiler_languages
```
I assume these two attributes are:
1. The list of all possible languages that this package may compile
2. The list of the languages provided by this package object (or better, by the associated concrete spec)

If so there are constraints among the two attributes that must be respected (2 is always $\subseteq$ 1), and they are not enforced when overriding. Wondering if we can do better :thinking:

becker33 replied:
Your assumption is correct. I can throw an assert in somewhere that `supported_languages` is a subset of `compiler_languages`.

becker33 replied:
I've created an `__init__` method that checks for that invariant.

alalazo replied:
Left another comment, so closing this.

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,126 @@
+    def supported_languages(self):
+        return self.compiler_languages
+
+    # TODO: how do these play nicely with other tags
+    tags = ["compiler"]
```
We might turn tags into a directive, similar to `maintainers`. We can do that in a separate PR.

alalazo replied:
Let's move this to a following PR.

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,126 @@
+            names.extend(getattr(cls, f"{language}_names"))
+        return names
+
+    @classproperty
+    def executables(cls):
+        regexp_fmt = r"^({0}){1}({2})$"
+        return [
+            regexp_fmt.format(prefix, re.escape(name), suffix)
+            for prefix, name, suffix in itertools.product(
+                cls.prefixes + [""], cls.compiler_names, cls.suffixes + [""]
+            )
+        ]
```
This is similar, but not equal, to:

https://github.com/spack/spack/blob/7e14ff806a63bf2af2a07f49fc13ed5b111a76ac/lib/spack/spack/compiler.py#L622-L641

Why did we remove the preamble? How many compilers do we support on Windows?

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,126 @@
+                return ".".join(match.groups())
+        except spack.util.executable.ProcessError:
+            pass
+        except Exception as e:
+            tty.debug(e)
```
We should probably leave a better trace on the provenance of the error, and not just dump the error message

alalazo replied:
Will commit a better message.

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,126 @@
+            tty.debug(e)
+
+    @classmethod
+    def determine_paths(cls, *, exes=None, prefix=None):
```
All these methods are worth a brief docstring. Here I assume the paths are related to compiler executables? Also, the name of the function is very generic, and if used as a mixin it won't be super clear which paths are determined. Maybe we should rename to something like `compiler_paths`?

becker33 replied:
Renamed to `determine_compiler_paths` and added docstrings to new methods.

alalazo reviewed (changes requested):
I checked the mixin and left a few comments. Overall I don't understand the technical reason why we can't split this in multiple PRs:
1. One to push common logic into the mixin
2. Others to add detection methods to other compilers
 
 Since this PR both refactors existing code _and_ changes the logic, it's more difficult to spot issues by just reviewing code.

That said, a very basic test seems to have regressed:
```
$ spack external find gcc
```
finds `gcc@9` in this PR on my laptop, and `gcc@9.4.0` on `develop`.

The newly added compilers:
- `aocc`
- `arm`
- `nag`

all miss a `detection_test.yaml` file to perform detection tests.

becker33 committed:
update intel compiler classic

becker33 committed:
update intel-oneapi-compilers package

becker33 committed:
Update compiler packages for nvhpc and pgi

becker33 committed:
update xl compiler package

combine packages for xlc and xlf

alalazo commented on var/spack/repos/builtin/packages/xlf/package.py:
```diff
@@ -1,69 +0,0 @@
-from spack.package import *
-
-
-class Xlf(Package):
-    """IBM XL Fortran is an advanced, high-performance compiler that can be
-    used for developing complex, computationally intensive programs, including
-    interlanguage calls with C and Fortran programs.
-    """
-
-    homepage = "https://www.ibm.com/support/knowledgecenter/SSXVZZ_16.1.1/com.ibm.compilers.linux.doc/welcome.html"
-
-    variant("r", default=True, description="The _r version of compilers")
-
-    def install(self, spec, prefix):
-        raise InstallError(
-            "XL compilers are not installable yet, but can be "
-            "detected on a system where they are supplied by vendor"
-        )
```
Why are we deleting this package and unifying with `xlc`? Older versions of `xlf` were higher than the corresponding versions of `xlc`, but I think the versions lined up at some point.  Are we no longer interested in supporting older versions with this mismatch ?

alalazo replied:
I guess we'll remove support for those old versions.

alalazo commented on var/spack/repos/builtin/packages/xl/detection_test.yaml:
```diff
@@ -5,6 +5,8 @@ paths:
           - "bin/xlc_r"
           - "bin/xlc++_r"
+          - "bin/xlf"
+          - "bin/xlf_r"
```
This is a wrong setup for the test, I don't think `xlf` will report:
```console
"IBM XL C/C++ for Linux, V16.1.1 (5725-C73, 5765-J13)"
```

alalazo commented on var/spack/repos/builtin/packages/xl/package.py:
```diff
@@ -27,43 +27,26 @@ def install(self, spec, prefix):
+    languages = ["c", "cxx", "fortran"]
+    c_names = ["xlc", "xlc_r"]
+    cxx_names = ["xlc++", "xlc++_r"]
+    fortran_names = ["xlf", "xlf_r"]  # TODO complete this
```
Why `xlC` is not among the names anymore?

alalazo commented on var/spack/repos/builtin/packages/xl/package.py:
```diff
@@ -27,43 +27,26 @@ def install(self, spec, prefix):
-            if "xlc++" in exe:
-                variants[variant_str]["cxx"] = exe
-                continue
+        _r_exes = [e for e in exes if "_r" in e]
+        _exes = [e for e in exes if "_r" not in e]
```
(minor) We can use `llnl.util.lang.stable_partition`

alalazo commented on var/spack/repos/builtin/packages/xl/package.py:
```diff
@@ -79,3 +62,10 @@ def cxx(self):
+    def fortran(self):
+        if self.spec.external:
+            return self.spec.extra_attributes["compilers"]["fortran"]
+        msg = "cannot retrieve C compiler [spec is not concrete]"
```
The message is reporting the wrong language (also the previous C++ message). Also, there's a mismatch between `compilers` and `paths` (I still prefer `compilers` because `paths` seems a bit generic on an external).

alalazo commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -503,11 +501,22 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
+    fortran_names = ["gfortran"]
+    d_names = ["gdc"]
+    go_names = ["gccgo"]
+    prefixes = [r"\w+-\w+-\w+-"]
+    suffixes = [r"-mp-\d+(?:\.\d+)?", r"-\d+(?:\.\d+)?", r"\d\d"]
```
Here we changed the suffixes and prefixes. Why?

becker33 replied:
The first suffix in the old code is included automatically by the `CompilerPackage` class.

The second suffix in the old code is covered by the first suffix here, but the regex is made more general to match the behavior from the `spack.compilers.gcc.Gcc`. 

The third and fourth suffixes in the old code, when combined, are identical to the second suffix here.

The fifth suffix in the old code is the fourth suffix here.

The prefixes here are added to better match the behavior of `spack.compilers.gcc.Gcc`.

alalazo replied:
> The first suffix in the old code is included automatically by the CompilerPackage class.

I'm not particularly fond of that, but let's postpone the discussion to another PR.

alalazo reviewed (dismissed):
I checked why this is reporting `gcc@9`. It's because we changed the flag being used from `-dumpfullversion` to `-dumpversion`: 
```console
$ gcc -dumpversion
9
$ gcc -dumpfullversion
9.4.0
```
It also seems we skip checking if:
```console
$ gcc --version
gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```
contains the `Apple` string to discard whether `gcc` is `apple-clang` in disguise?

Can we please revert that? 

I still think this would be much easier to review if split in smaller PRs doing a single change, rather than a big PR that overhauls code and regexes and flags together.

I'll submit a separate change to the detection test that makes this fail, so we also test which flag is supposed to be used.

alalazo commented on lib/spack/spack/test/compilers/detection.py:
```diff
@@ -24,6 +25,51 @@
+from spack.pkg.builtin.xl import Xl
+
+
+def check_package_detection(mock_executable, output, expected_version, cls):
+    languages = cls.compiler_languages
+
+    executables = {}
+    script = "\n".join(f'echo "{line}"' for line in output.split("\n"))
+    for lang in languages:
+        name = getattr(cls, f"{lang}_names")[0]
+        executables[name] = mock_executable(name, output=script)
+        bindir = os.path.dirname(executables[name])
+        with open(executables[name], "r") as f:
+            print(f.read())
+
+    detected = spack.detection.by_path(
+        [cls.fullname.replace("_", "-")], path_hints=[bindir], max_workers=1
+    )
+
+    spec_name = cls.name.replace("_", "-")
+    assert len(detected) == 1
+    assert spec_name in detected
+
+    detected_package = detected[spec_name]
+    assert len(detected_package) == 1
+    assert detected_package[0].spec.satisfies(f"{spec_name}@{expected_version}")
+    assert detected_package[0].prefix == os.path.dirname(bindir)
+
+    suffix = ".bat" if sys.platform == "win32" else ""
+    for lang in languages:
+        name = getattr(cls, f"{lang}_names")[0]
+        assert detected_package[0].spec.extra_attributes["paths"][lang] == os.path.join(
+            bindir, name + suffix
+        )
```
Why are we adding this? We already have `detection_tests.yaml` to check package detection. I don't get why we are duplicating the test... :thinking:

becker33 replied:
`detection_tests.yaml` runs on a single architecture. I've already caught bugs in this PR that wouldn't have been caught by `detection_tests.yaml`.

alalazo replied:
Which bugs? Why they would not be caught?

becker33 replied:
Windows-specific bugs -- we don't run audits on Windows. Also some bugs that could have been expressed in an audit, but which weren't included (for a package that already has audits) and would've otherwise slipped through unless I happened to think to add them.

alalazo replied:
> Windows-specific bugs -- we don't run audits on Windows.

Why not adding audits on Windows then? 

> Also some bugs that could have been expressed in an audit [...]

I'm not sure I understand what you mean here.

alalazo replied:
Windows audits have been added, and the ability to check extra attributes too. I'll try to remove this extra code.

alalazo replied:
Extra code removed.

alalazo commented on var/spack/repos/builtin/packages/arm/package.py:
```diff
@@ -0,0 +1,28 @@
+
+    version("12.0.0")
+
+    def install(self, spec, prefix):
+        raise NotImplementedError("apple-clang compiler must be configured as external")
```
Message above needs update

becker33 replied:
Relevant code was removed

becker33 committed:
merge new arm package into existing acfl package

becker33 committed:
fix gcc to use -dumpfullversion when necessary

becker33 committed:
refactor to add documentation and make naming clearer

becker33 committed:
avoid class-scoped lambda

becker33 committed:
add guardrails for supported_languages <= compiler_languages

becker33 committed:
re-add missing windows support in executables computation

becker33 committed:
resolve test failures from rebase

becker33 committed:
add cce package

becker33 committed:
fixup 4787b5df64   update test to match

becker33 committed:
update test to match new syntax

becker33 committed:
do not test case-sensitive compilers on windows

becker33 committed:
update llvm compiler package

becker33 committed:
update llvm-amdgpu (rocmcc) compiler package

becker33 committed:
add msvc package mapping to msvc compiler

becker33 committed:
debug windows test failure

becker33 committed:
fixup typo in 85c5c212d8

becker33 committed:
fixup file layout in xl detection_test.yaml

becker33 committed:
Add xlC in addition to xlc++ in Xl.cxx_names

becker33 committed:
llvm-amdgpu: fixup imports

becker33 committed:
more debugging

becker33 committed:
no quote around echo on windows

becker33 committed:
style

becker33 committed:
remove debug print statements

haampie commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,154 @@
+    # metadata identifying this as a compiler for lmod
+    family = "compiler"
+
+    # TODO: how do these play nicely with other tags
```
similar to `maintainers = [*xs]` -> `maintainers(*xs)`

alalazo replied:
We should submit a small PR doing that, either as a pre-requisite for this one, or immediately after. See the comment on `llvm`

alalazo replied:
Let's move this to a separate PR.

haampie commented on lib/spack/spack/package_base.py:
```diff
@@ -249,6 +249,7 @@ def determine_spec_details(cls, prefix, objs_in_prefix):
                 specs = []
                 for version_str, objs in objs_by_version.items():
                     variants = cls.determine_variants(objs, version_str)
+
```
```suggestion
```

alalazo commented on var/spack/repos/builtin/packages/llvm/package.py:
```diff
@@ -8,13 +8,14 @@
 
 
-class Llvm(CMakePackage, CudaPackage):
+class Llvm(CMakePackage, CudaPackage, CompilerPackage):
```
`llvm` has tags of its own, therefore it's not detected as a compiler when we use:
```console
$ spack external find --tag compiler
```

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,154 @@
+
+
+class CompilerPackage(spack.package_base.PackageBase):
+    """A Package mixin for all common logic for packages that implement compilers"""
```
(nitpick) Maybe let's make it clear that right now this is about detection:
```suggestion
    """Mixin with common logic used when detecting compilers"""
```

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,154 @@
+
+    # Optional suffix regexes for searching for this type of compiler.
+    # Suffixes are used by some frameworks, e.g. macports uses an '-mp-X.Y'
+    # version suffix for gcc.
```
```suggestion
    #: Optional suffix regexes for searching for this type of compiler.
    #: Suffixes are used by some frameworks, e.g. macports uses an '-mp-X.Y'
    #: version suffix for gcc.
```

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,154 @@
+    # version suffix for gcc.
+    suffixes: Sequence[str] = [r"-.*"]
+
+    # Optional prefix regexes for searching for this compiler
```
```suggestion
    #: Optional prefix regexes for searching for this compiler
```

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,154 @@
+    #: Compiler argument(s) that produces version information
+    #: If multiple arguments, the earlier arguments must produce errors when invalid
+    #: Must be a hashable type (e.g. tuple, rather than list)
+    version_argument = "-dumpversion"
```
Maybe let's just use:
```suggestion
    version_arguments: List[str] = ["-dumpversion"]
```
so that the type is clear.

alalazo replied:
Closing this one, new comment in https://github.com/spack/spack/pull/43464/files#r1574929065

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,154 @@
+        if isinstance(version_argument, str):
+            version_argument = [version_argument]
+
+        for va in version_argument:
```
We can then remove this:
```suggestion
        for va in cls.version_arguments:
```

alalazo replied:
Let's also postpone the discussion on this. I still like a clear type better, but this is not a major issue.

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,154 @@
+    version_argument = "-dumpversion"
+
+    #: Return values to ignore when invoking the compiler to get its version
+    ignore_version_errors: Sequence[int] = ()
```
I don't see where this is used. We should either replicate what compiler objects do, and add it in the call to:
```
get_compiler_version_output
```
below, or remove it.

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,154 @@
+    ignore_version_errors: Sequence[int] = ()
+
+    #: Regex used to extract version from compiler's output
+    version_regex = "(.*)"
```
```suggestion
    version_regex: str = "(.*)"
```

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,154 @@
+
+    #: Platform matcher for Platform objects supported by compiler
+    # The value must be callable
+    is_supported_on_platform = true
```
I'd rather do:
```python
@staticmethod
def is_supported_on_platform(platform) -> bool:
    return True
```
and delete the the function above. Besides that, this function is currently not used (unless I'm missing something) so I think we should remove it altogether.

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,154 @@
+    is_supported_on_platform = true
+
+    #: Static definition of languages supported by this class
+    compiler_languages = ["c", "cxx", "fortran"]
```
```suggestion
    compiler_languages: List[str] = ["c", "cxx", "fortran"]
```

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,154 @@
+    #: Dynamic definition of languages supported by this package
+    @property
+    def supported_languages(self):
+        return self.compiler_languages
```
```suggestion    
    @property
    def supported_languages(self):
        """Dynamic definition of languages supported by this package"""
        return self.compiler_languages
```

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,154 @@
+    def executables(cls):
+        """Construct executables for external detection from names, prefixes, and suffixes."""
+        regexp_fmt = r"^({0}){1}({2})$"
+        prefixes = [""] + cls.prefixes
+        suffixes = [""] + cls.suffixes
```
Why do we want to hard-code the case with no prefixes/suffixes?

becker33 replied:
Because the packages shouldn't have to each individually specify that the executable names can appear without any prefix/suffix.

alalazo replied:
Is there any case we know where we want to get rid of `""` ? Right now it would be impossible.

alalazo replied:
See above, let's postpone the discussion of this detail.

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,154 @@
+        """Compute the paths to compiler executables associated with this package
+
+        This is a helper method for ``determine_variants`` to compute the ``extra_attributes``
+        to include with each spec object."""
```
Hmm, not sure I like this API, but we might postpone discussion on this if we don't find an agreement. Basically:
1. The arguments are switches on behavior (either compute from exes, or guess exes from prefix)
2. Computing from prefixes is used during auto-detection, which is what masked the bug we discovered here (using the wrong dictionary entries, i.e. `paths` instead of `compilers` in the previous PR)

Would it be possible to just remove auto-detection and document how a package entry should be written to be interpreted as a compiler?

alalazo replied:
I added https://github.com/spack/spack/pull/43464/commits/805f78769694f4fddf464b97dad838b545d92ba7 which has a comment open below, so let's discuss there.

alalazo commented on lib/spack/spack/compilers/__init__.py:
```diff
@@ -173,25 +173,34 @@ def _compiler_config_from_external(config):
+    attribute_paths = extra_attributes.get("compilers", {})
+    supported_languages = set(
+        getattr(spec.package_class(spec), "supported_languages", ("c", "cxx", "fortran"))
+    )
```
In which case a compiler does not have a `supported_languages` attribute? Shouldn't we just record a failure in a debug message and skip that compiler?

alalazo replied:
Outdated

alalazo commented on lib/spack/spack/compilers/__init__.py:
```diff
@@ -173,25 +173,34 @@ def _compiler_config_from_external(config):
+        getattr(spec.package_class(spec), "supported_languages", ("c", "cxx", "fortran"))
+    )
+    if prefix and not supported_languages.issubset(set(attribute_paths.keys())):
+        paths = spec.package_class.determine_compiler_paths(prefix=prefix)
```
If a supported language is not in a `packages.yaml` entry, wouldn't it just be easier to leave a log message and move on to the next compiler?

alalazo replied:
Outdated

alalazo commented on lib/spack/spack/compilers/__init__.py:
```diff
@@ -173,25 +173,34 @@ def _compiler_config_from_external(config):
+    paths.update(attribute_paths)
+
+    # compilers format has cc/fc/f77, externals format has "c/fortran"
+    if "c" in paths:
+        paths["cc"] = paths.pop("c")
+    if "fortran" in paths:
+        fc = paths.pop("fortran")
+        paths["fc"] = fc
+        if "f77" not in paths:
+            paths["f77"] = fc
```
Took me a while to read this. Why not using 2 dictionaries, instead of popping and assigning keys in place?

alalazo commented on lib/spack/spack/test/compilers/detection.py:
```diff
@@ -24,6 +25,54 @@
+    )
+
+    spec_name = cls.name.replace("_", "-")
+    print(detected)
+    print(detected[spec_name][0])
+    print(detected[spec_name][0].spec.extra_attributes)
```
:thinking:

alalazo commented on lib/spack/spack/test/compilers/detection.py:
```diff
@@ -4,6 +4,7 @@
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 """Test detection of compiler version"""
 import os
+import sys
```
In general, I think these tests should not use `check_package_detection`, but we should instead:
1. Write `detection_tests.yaml` for each compiler
2. Add the features that are currently missing to `spack audit externals` (which are them?)

Here we are importing a lot of `builtin` packages, which couples the tests with the `builtin` repository, and is not scalable. In my opinion one objective would be to have packages that can be used as a compiler and carry their own tests wherever repository they are located in.

alalazo replied:
Done

alalazo commented on var/spack/repos/builtin/packages/apple-clang/package.py:
```diff
@@ -2,57 +2,23 @@
+    cxx_names = ["clang++"]
 
-        return "", {"compilers": compilers}
+    version_regex = r"^Apple (?:LLVM|clang) version ([^ )]+)"
```
Here we are missing the multiline option. Not sure it matters for recent Apple Clang.

becker33 replied:
Is there any reason to keep the `^`? If we get rid of that then multiline becomes completely irrelevant (although it's also irrelevant for every apple-clang version output I've been able to compare to).

alalazo replied:
Tests are passing, so let's close this discussion. We can get back to it if we see faulty detection.

alalazo commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -503,11 +501,37 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
+
+    @property
+    def supported_languages(self):
+        # This weirdness is because it could be called on an abstract spec
+        if "languages" not in self.spec.variants:
+            return self.compiler_languages
+        return [x for x in self.compiler_languages if x in self.spec.variants["languages"].value]
```
Not sure how to improve this at the moment :thinking:

alalazo commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -503,11 +501,37 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
+        except Exception:
+            output = ""
+        # Apple's gcc is actually apple clang, so skip it.
+        # Users can add it manually to compilers.yaml at their own risk.
+        if "Apple" in output:
```
```suggestion
        if "Apple" in output:
```

alalazo reviewed (commented):
From a user perspective, the most visible issue is that `llvm` is not detected, because the `tag` attribute from the mixin is overridden. This can be fixed quite easily.

I left mostly minor comments on code, but there are a couple of major points:
1. I would avoid extensive use of auto-detection for entries in packages.yaml
2. I would prefer to have detection tests alongside packages instead of in a centralized place

Which features are missing in `spack audit external` to get point 2 ?

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,154 @@
+    compiler_languages = ["c", "cxx", "fortran"]
+
+    def __init__(self, spec):
+        super().__init__(spec)
+        assert set(self.supported_languages) <= set(self.compiler_languages)
```
Where do we expect this to possibly fail? I am wondering if there are cases where this would make Spack unusable in cases where it could indeed proceed (similar to the issue with `dpcpp` being an error instead of a warning)

becker33 replied:
The only way for this to fail is when you write a malformed `supported_languages` property and then also write a bad external entry for that package.

alalazo replied:
If we keep the exception, can we raise a better one with a message? Currently if a user stumble on this for any reason the message would be `AssertionError:`

alalazo replied:
Still `AssertionError`, but there's a much improved message. Closing the comment.

becker33 committed:
remove vestigial line change

becker33 committed:
update llvm tag

becker33 committed:
CompilerPackage: remove vestigial 'family' attribute

becker33 committed:
clean up CompilerPackage interface/documentation

becker33 committed:
CompilerPackage: clean up interface to get compiler paths

becker33 committed:
simplify compiler config conversion logic

becker33 committed:
remove comment per review

becker33 committed:
unify version detection between acfl and arm compiler

becker33 committed:
cce: improve version regex

becker33 committed:
fj/aocc/cce/acfl detection_test.yaml

becker33 committed:
fixup msvc detection

becker33 committed:
external audits for the rest of the compiler packages

becker33 committed:
remove vestigial import

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,145 @@
+
+    #: Compiler argument(s) that produces version information
+    #: If multiple arguments, the earlier arguments must produce errors when invalid
+    version_argument: Union[str, Tuple[str]] = "-dumpversion"
```
What do we gain by also allowing a string here? I don't expect people to write version argument that often to be bothered by `[]`, so I would still advocate for not using a union type, and simplify the logic later.

alalazo replied:
Minor issue that shouldn't be a blocker, though I prefer a single type.

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,145 @@
+    #: Optional suffix regexes for searching for this type of compiler.
+    #: Suffixes are used by some frameworks, e.g. macports uses an '-mp-X.Y'
+    #: version suffix for gcc.
+    suffixes: Sequence[str] = [r"-.*"]
```
We should probably add a prefix to names that are too generic for a mixin:
```suggestion
    compiler_suffixes: Sequence[str] = [r"-.*"]
```

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,145 @@
+    suffixes: Sequence[str] = [r"-.*"]
+
+    #: Optional prefix regexes for searching for this compiler
+    prefixes: Sequence[str] = []
```
```suggestion
    compiler_prefixes: Sequence[str] = []
```

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,145 @@
+    version_argument: Union[str, Tuple[str]] = "-dumpversion"
+
+    #: Regex used to extract version from compiler's output
+    version_regex: str = "(.*)"
```
```suggestion
    compiler_version_regex: str = "(.*)"
```

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,145 @@
+        return names
+
+    @classproperty
+    def executables(cls):
```
```suggestion
    def compiler_executables(cls):
```

becker33 replied:
This one has to be `executables` because that's the name the detection logic relies on -- I can document that better.

alalazo replied:
Apologies, I overlooked that one then.

alalazo commented on lib/spack/spack/build_systems/compiler.py:
```diff
@@ -0,0 +1,145 @@
+        return paths
+
+    @classmethod
+    def determine_variants(cls, exes, version_str):
```
Can you add type hints where missing in this file?

alalazo reviewed (commented):
A few more minor comments. I'll start working on allowing audits per platform

becker33 committed:
CompilerPackage: type hints and better names

becker33 committed:
update packages to match interface change

alalazo committed:
acfl: specify platforms for external detection

alalazo committed:
aocc: specify platforms for external detection

alalazo committed:
cce: specify platforms for external detection

alalazo committed:
fj: specify platforms for external detection

alalazo committed:
msvc: specify platforms for external detection

alalazo committed:
nag: specify platforms for external detection

alalazo committed:
nvhpc: specify platforms for external detection

alalazo committed:
pgi: specify platforms for external detection

alalazo committed:
oneapi: specify platforms for external detection

alalazo committed:
intel: specify platforms for external detection

alalazo committed:
llvm-amdgpu: specify platforms for external detection

alalazo committed:
Improve debug message for failing version detection

alalazo committed:
Restore compiler unit tests to previous state

alalazo committed:
Add checks for extra attributes

alalazo committed:
Format YAML detection homogeneously

alalazo committed:
Update cce detection test

alalazo committed:
Do not auto-detect missing entries from packages.yaml

Instead, leave either a debug message or a loud warning,
depending on the entry being missing.

Delete a unit test that was detecting `gcc-8` as being
part of a `gcc@7.7.7` spec, with previous auto-detection.

alalazo commented on lib/spack/spack/test/cmd/compiler.py:
```diff
@@ -285,26 +284,7 @@ def test_compiler_list_empty(no_compilers_yaml, working_env, compilers_dir):
-            },
-            """gcc@7.7.7:
-\tpaths:
-\t\tcc = {compilers_dir}{sep}gcc-8{suffix}
-\t\tcxx = {compilers_dir}{sep}g++-8{suffix}
-\t\tf77 = {compilers_dir}{sep}gfortran-8{suffix}
-\t\tfc = {compilers_dir}{sep}gfortran-8{suffix}
```
I deleted this one, cause detecting `gcc-8` for a `gcc@7.7.7` spec seems wrong. I think it's better to warn/leave a debug message and avoid auto-detection.

alalazo replied:
See https://github.com/spack/spack/pull/43464/commits/805f78769694f4fddf464b97dad838b545d92ba7

alalazo commented:
Relevant changes that we should note:
1. `xlf` and `xlc` have been merged in a unique `xl` package. That prevent us from being able to detect old versions of the compiler suite, where `xlf` was 2 major version ahead of `xlc`.
2. We changed a regex, see https://github.com/spack/spack/pull/43464#discussion_r1571116055 I'm leaving a trace here in case we come back to this PR later
3. Auto-detection of compiler prefixes, when converting a packages.yaml entry to a compilers.yaml entry, have been removed. Missing attributes now would print a warning or a debug message to users.

alalazo commented:
@becker33 Let me know if there's anything in the latest commits you don't agree with.

amd-toolchain-support commented:
Testing for AOCC support I have noticed that `spack external find {aocc,gcc,...}` will detect and re-add Spack provided compilers if they are loaded at the point of searching.  Is this intended behaviour?

```
$ spack install aocc@4.2+license-agreed
...
$ spack load aocc
$ spack external find aocc
==> The following specs have been detected on this system and added to /home/$USER/repos/spack/etc/spack/packages.yaml
aocc@4.2.0
```

Otherwise AOCC detection is behaving as I would expect.

alalazo commented:
> Is this intended behaviour?

Yes, this is how external detection for packages work at the moment. It's just based on PATH inspection.

alalazo commented:
(To be solved in a following PR) There is still one thing that is missing for:
```console
spack external find -t compiler
```
to be a full substitute of `spack compiler find`, i.e. we need some way to have an equivalent of:
```console
spack compiler find --mixed-toolchain
```
